core/src/main/java/nl/nn/adapterframework/filesystem/ExchangeFileSystem.java

```diff
@@ -253,19 +211,96 @@ public FolderId findFolder(FolderId baseFolderId, String folderName) throws Exce
 		return result;
 	}
 
+	private ExchangeService createExchangeService() throws FileSystemException {
+		ExchangeService exchangeService = new ExchangeService(ExchangeVersion.Exchange2010_SP2);
+		
+		String defaultUsername = StringUtils.isEmpty(getAccessToken())? getUsername() : null;
+		String defaultPassword = StringUtils.isEmpty(getAccessToken())? getPassword() : getAccessToken();
+
+		CredentialFactory cf = new CredentialFactory(getAuthAlias(), defaultUsername, defaultPassword);
+		if (StringUtils.isEmpty(cf.getUsername())) {
+			// use OAuth Bearer token authentication
+			exchangeService.getHttpHeaders().put("Authorization", "Bearer "+cf.getPassword());
+		} else {
+			// use deprecated Basic Authentication. Support will end 2021-Q3!
+			log.warn("Using deprecated Basic Authentication method for authentication to Exchange Web Services");
+			ExchangeCredentials credentials = new WebCredentials(cf.getUsername(), cf.getPassword());
+			exchangeService.setCredentials(credentials);
+		}
+
+		
+		
+		if (StringUtils.isNotEmpty(getProxyHost()) && (StringUtils.isNotEmpty(getProxyAuthAlias()) || StringUtils.isNotEmpty(getProxyUsername()) || StringUtils.isNotEmpty(getProxyPassword()))) {
+			CredentialFactory proxyCf = new CredentialFactory(getProxyAuthAlias(), getProxyUsername(), getProxyPassword());
+			WebProxyCredentials webProxyCredentials = new WebProxyCredentials(proxyCf.getUsername(), proxyCf.getPassword(), getProxyDomain());
+			WebProxy webProxy = new WebProxy(getProxyHost(), getProxyPort(), webProxyCredentials);
+			exchangeService.setWebProxy(webProxy);
+		}
+
+		RedirectionUrlCallback redirectionUrlCallback = new RedirectionUrlCallback() {
+
+			@Override
+			public boolean autodiscoverRedirectionUrlValidationCallback(String redirectionUrl) {
+				if (isValidateAllRedirectUrls()) {
+					log.debug("validated redirection url ["+redirectionUrl+"]");
+					return true;
+				}
+				log.debug("did not validate redirection url ["+redirectionUrl+"]");
+				return super.autodiscoverRedirectionUrlValidationCallback(redirectionUrl);
+			}
+			
+		};
+
+		if (StringUtils.isEmpty(getUrl())) {
+			log.debug("performing autodiscovery for ["+getMailAddress()+"]");
+			try {
+				exchangeService.autodiscoverUrl(getMailAddress(),redirectionUrlCallback);
+			} catch (Exception e) {
+				throw new FileSystemException("cannot autodiscover for ["+getMailAddress()+"]", e);
+			}
+		} else {
+			try {
+				exchangeService.setUrl(new URI(getUrl()));
+			} catch (URISyntaxException e) {
+				throw new FileSystemException("cannot set URL ["+getUrl()+"]", e);
+			}
+		}
+		log.debug("using url ["+exchangeService.getUrl()+"]");
+		return exchangeService;
+	}
+	
+	private ExchangeService getExchangeService() {
+		ExchangeService localExchangeService = threadLocalExchangeService.get();
+		return localExchangeService !=null ? localExchangeService : globalExchangeService;
```

Comment:
Ik snap hier niets van. Waarom geef je niet altijd terug wat in de threadlocal zit. Ik zie iig geen toegevoegde waarde om het naast een threadlocal ook in een global bij te houden...
De threadlocal moet/kan een `initialValue` hebben waar hij de `ExchangeService` aanmaakt. Dan hoef je geen `return localExchangeService !=null ? localExchangeService : createService();` te doen...

Reply:
In openThread() en closeThread() worden de threadLocal variabelen gemanaged voor de FileSystemListener, die als PullingListener in een Receiver kan draaien. Dan krijgt elke thread z'n eigen ExchangeService. 
Bij gebruik als ExchangeFolderSender is er geen openThread()/closeThread(), maar je/ik wil ook niet dat alle 100 WebSphere threads een eigen ExchangeService gaan hebben, en ook niet dat die voor elk verzoek een nieuwe connectie moet maken. 
Misschien moet er voor dat geval een 'releaseConnection()' komen.